File: DailyChallenges/Scala/LeetCode_904.scala
object Solution {
    def totalFruit(fruits: Array[Int]): Int = {
        val cnt = collection.mutable.HashMap.empty[Int,Int]
        var ans = 0; var l = 0; var r = 0
        while (r < fruits.length){
            cnt(fruits(r)) = cnt.getOrElse(fruits(r), 0) + 1
            while (cnt.size > 2){
                cnt(fruits(l)) = cnt(fruits(l)) - 1
                if (cnt(fruits(l)) == 0)
                    cnt.remove(fruits(l))
                l += 1
            }
            ans = Math.max(ans, r-l+1)
            r += 1
        }
        ans
    }
}

File: DailyChallenges/CPP/LeetCode_904.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int> cnt;
        int ans = 0, l = 0, r = 0;
        while (r < fruits.size()){
            ++cnt[fruits[r++]];
            while(cnt.size() > 2) {
                if (--cnt[fruits[l]] == 0)
                    cnt.erase(fruits[l]);
                ++l;
            }
            ans = max(ans, r-l);
        }
        return ans;
    }
};
